contracts/access/AccessControl.sol

```diff
@@ -0,0 +1,178 @@
+pragma solidity ^0.6.0;
+
+import "../utils/EnumerableSet.sol";
+import "../GSN/Context.sol";
+
+/**
+ * @dev Contract module that allows children to implement role-based access
+ * control mechanisms.
+ *
+ * Roles are referred to by their `bytes32` identifier. These should be exposed
+ * in the external API and be unique. The best way to achieve this is by
+ * using `public constant` hash digests:
+ *
+ * ```
+ * bytes32 public constant MY_ROLE = keccak256("MY_ROLE");
+ * ```
+ *
+ * Roles can be used to represent a set of permissions. To restrict access to a
+ * function call, use {hasRole}:
+ *
+ * ```
+ * function foo() public {
+ *     require(hasRole(MY_ROLE, _msgSender()));
+ *     ...
+ * }
+ * ```
+ *
+ * Roles can be granted and revoked programatically by calling the `internal`
+ * {_grantRole} and {_revokeRole} functions.
+ *
+ * This can also be achieved dynamically via the `external` {grantRole} and
+ * {revokeRole} functions. Each role has an associated admin role, and only
+ * accounts that have a role's admin role can call {grantRole} and {revokeRoke}.
+ *
+ * By default, the admin role for all roles is `DEFAULT_ADMIN_ROLE`, which means
+ * that only accounts with this role will be able to grant or revoke other
+ * roles. More complex role relationships can be created by using
+ * {_setRoleAdmin}.
+ */
+abstract contract AccessControl is Context {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    struct Role {
+        EnumerableSet.AddressSet members;
+        bytes32 admin; // This role's admin role
+    }
+
+    mapping (bytes32 => Role) private _roles;
+
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+
+    /**
+     * @dev Emitted when `account` is granted `roleId` role.
+     */
+    event RoleGranted(bytes32 indexed roleId, address indexed account);
```

Comment:
We should also log the address of the admin that did the granting. It's not easy to get this information otherwise.

Reply:
We've had similar discussions as to what that address should be on programmatic interactions when talking about ERC777 operators.

What do you think about the following? (pseudo code)

```solidity
function grantRole(role, recipient) external {
    checkAdminRole(role, msg.sender);
    _grantRole(role, recipient, msg.sender);
}

function grantRole(role, recipient) internal {
    _grantRole(role, recipient, address(this));
}

function _grantRole(role, recipient, operator) private {
    roles[role].operator = true;
    emit RoleGranted(role, recipient, operator);
}
```

Reply:
My feeling is it is more useful to always log `msg.sender`. Even if the granting was done through logic of the contract, I think for auditing/monitoring purposes it's more relevant to know who triggered it.

Reply:
I have this same intuition for ERC777 and all other similar cases and I would really consider changing this in 3.0.

Reply:
That makes sense. I'll address this.